### PR TITLE
kernel: add makefile entry for IXGBE SRIOV VF

### PR DESCRIPTION
**Pull request by cblakely**
Building off PR #302 / #313 - including ixgbe SR-IOV virtual function network driver. Tested against SR-IOV VF on Intel X520 as KVM guest (built with target x86_64).

### CODE REVIEW

**Comment by pprindeville** [comment]
No issues with the patch itself, but...  why is this needed?

**Comment by cblakely** [comment]
SRIOV Virtual Function is basically a hardware switching function in the NIC silicon, rather than using something like open-vswitch with virtio, or linux bridging on your hypervisor / compute hosts. It allows lower overhead and greater performance when compared to virtualized network interfaces.

I realize this is pretty far from something you would see in a hardware router device - it really only applies to using OpenWRT as a virtual machine on compute hardware with Intel 10G 82599 network adapters, such as what we use for our compute nodes to provide virtual networking in some labs. OpenWRT makes a great edge firewall appliance with enough API capability to script basic operations against.

I've been patching it and producing custom builds for a while for both IXGBE and IXGBEVF. I saw that IXGBE was recently added to master so I wanted to add the VF driver as well, which would eliminate the need for locally patching builds. Also you already support VF for Intel IGB (1g interfaces).

**Comment by pprindeville** [comment]
> SRIOV Virtual Function is basically a hardware switching function in the NIC silicon, rather than using something like open-vswitch with virtio, or linux bridging on your hypervisor / compute hosts. It allows lower overhead and greater performance when compared to virtualized network interfaces.

No, I know what SR-IOV is.  I spent a year at Intel Labs working on Xeon architecture.  And indeed, I run this on my Broadwell virtualization server (Centos 6.8 host).

The functionality is useful for Xen/KVM as the host OS.  It's not exposed to the guest OS, which I think is what you're proposing.

Unless you're saying you want to run KVM on OpenWRT and run guest OS's hosted by OpenWRT?

I would be inclined to go the opposite direction: create a stable guest OS platform w/ SR-IOV (exactly like my Centos virtualization server) and then:

1. isolate the CentOS host from outside access, but
2. run OpenWRT (or pfsense or whatever) as the firewall in a guest VM

that way, even if your firewall is compromised (hey, it happens) then it at least is sandboxed.

You could also do some sort of health attestation of the guest VM, and if it fails, shutdown the firewall VM.

**Comment by cblakely** [comment]
Oh, sorry if I wasn't clear - definitely openwrt is the VM. The issue is that SRIOV VF manifests in the guest as a raw PCI device, and needs that kernel module / driver. The host (ie linux/KVM on the bare metal compute host) needs the IXGBE driver which is used to provision the VFs, but the host itself doesn't interact with the VF. So current master branch has IXGBE support which great for bare metal, but without the VF driver Openwrt doesn't use the SRIOV PCI device inside the VM.

**Comment by cblakely** [comment]
define PF network pool on hypervisor:
```
vmhost1:~# virsh net-dumpxml ens1f1-sriov-pool
<network>
  <name>ens1f1-sriov-pool</name>
  <uuid>8c904a71-c920-4c64-bc52-68426b3d6f03</uuid>
  <forward mode='hostdev' managed='yes'>
    <pf dev='ens1f1'/>
  </forward>
</network>
```

excerpt from libvirtd domain.xml on openwrt guest for VF interface:
```
    <interface type='network'>
      <mac address='52:54:00:cd:7c:4c'/>
      <source network='ens1f1-sriov-pool'/>
      <vlan>
        <tag id='10'/>
      </vlan>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
    </interface>
```

VF driver attaching from within OpenWRT running as VM guest
```
root@OpenWrt:~# dmesg | grep ixgbe
[    4.692334] ixgbe: Intel(R) 10 Gigabit PCI Express Network Driver - version 4.2.1-k
[    4.693466] ixgbe: Copyright (c) 1999-2015 Intel Corporation.
[    4.694560] ixgbevf: Intel(R) 10 Gigabit PCI Express Virtual Function Network Driver - version 2.12.1-k
[    4.696657] ixgbevf: Copyright (c) 2009 - 2012 Intel Corporation.
[    4.712090] ixgbevf 0000:00:03.0: 52:54:00:cd:7c:4c
[    4.712801] ixgbevf 0000:00:03.0: MAC: 1
[    4.713411] ixgbevf 0000:00:03.0: Intel(R) 82599 Virtual Function
[    5.897293] ixgbevf 0000:00:03.0: NIC Link is Up 10 Gbps
```

**Comment by blogic** [comment]
patch is missing a proper subject prefix and Signed-off-by line. this is required so that we can merge the fix into LEDE

**Comment by cblakely** [comment]
updated commit based on your guidelines here: https://lede-project.org/submitting-patches

File: package/kernel/linux/modules/netdevices.mk [hunk 1]
```diff
@@ -523,6 +523,7 @@ endef
 
 $(eval $(call KernelPackage,igb))
 
+
```
Comment (pprindeville) [inline package/kernel/linux/modules/netdevices.mk]: Unnecessary whitespace.

**Review by pprindeville — CHANGES_REQUESTED** [review]
Please squash your commits (actually fix using "git rebase -i HEAD~2") and force push.

File: package/kernel/linux/modules/netdevices.mk (hunk 1 above)
Comment (cblakely) [reply package/kernel/linux/modules/netdevices.mk]: Whitespace followed the conventions already in the file - two empty lines before each "define" block

**Comment by cblakely** [comment]
Squashed commits

File: package/kernel/linux/modules/netdevices.mk (hunk 1 above)
Comment (pprindeville) [reply package/kernel/linux/modules/netdevices.mk]: That may be, but each commit also needs to be singular in purpose and focus.  We don't hit "targets of opportunity" because that makes backing out a bad commit too problematic.

In general, use the "light touch" when working on the code base and you can't go wrong.

**Review by pprindeville — CHANGES_REQUESTED** [review]

**Comment by blogic** [comment]
@cblakely what @pprindeville is trying to say in a posh way is "one change per commit please"

**Comment by blogic** [comment]
merged a version with minor changes and a proper description into my staging tree, Thanks u

**Comment by pprindeville** [comment]
Posh, wow. Never been accused of being posh before. Redneck and yobo, sure...  I’ll add it to my list of Appellations.

**Comment by cblakely** [comment]
I am in total agreement with the "one change per pull" mantra. Whitespace just didn't seem enough like a change for me to worry about it. That said it's your codebase 👍 

Anyway, thanks for merging, that will be great!

**Comment by pprindeville** [comment]
If you're interested, here's the rationale:

suppose I commit a fix with whitespace changes... then someone else commits another patch which overlaps it... there will be a conflict.  so they rebase their patch and commit it.

then our patch is found to be defective, so we revert it.

but the 2nd fix that someone else did, now needs to be cherry-picked into 17.01.  but it won't go, because it won't match the region with the whitespace changes (which never got cherry-picked to 17.01).

so it needs to be rebased and hand-edited.

just a lot of trouble for nothing.
